Chess

tootiltedboy opened the pull request:
Написана базовая структура для консольного приложения, игры в шахматы.
На данный момент есть: 
- Поддержка unicode символов для отображения на игровой доске
- Проверка правильности ввода хода
- Поддержка игровых правил (за исключением рокировки)
- Проверка шаха и мата

tootiltedboy committed:
Made the in-console chess game basic assets, the unicode symbols support and basic game rules check were also added

tootiltedboy committed:
Made the in-console chess game basic assets, the unicode symbols support and basic game rules check were also added

tootiltedboy committed:
Added CMakeList.txt to provide the compilation of the project

trimanul commented:
Сейчас пешка не может сделать ход на 2 клетки вперед из стартовой позиции

trimanul commented:
Надо доработать отрисовку доски, т.к. сейчас это немного нечитаемо(возможно добавить пробелов у фигур, или устанавливать определенный размер шрфита). Если это ток на моем линукс терминале, то претензий не имею тогда
![изображение](https://github.com/user-attachments/assets/ba1ed99a-1185-446f-817a-acdeafd80b7d)

trimanul commented on exercises/Chess-build/main.cpp:
```diff
@@ -0,0 +1,132 @@
+    ChessBoard chessBoard;
+    string currentPlayer = "White";
+
+    while (true) {
+        chessBoard.printBoard();
+        cout << currentPlayer << "'s turn. Enter your move (e.g., KI E2 E4): ";
+        string move;
+        getline(cin, move);
+
+        if (chessBoard.movePiece(move)) {
+            currentPlayer = (currentPlayer == "White") ? "Black" : "White";
+        }
+    }
```
Нигде не вижу кода для остановки цикла. Игра будет идти бесконечно, т.к. даже если король в шахе, и у него нет возможности из него выйти(aka мат), то movePiece будет просто бесконечно возвращать false. Нужен кондишн остановки игры

trimanul commented on exercises/Chess-build/main.cpp:
```diff
@@ -0,0 +1,132 @@
+        Piece king = (piece < BLACK_PAWN) ? WHITE_KING : BLACK_KING;
+        int kingRow, kingCol;
+
+        for (int i = 0; i < 8; i++) {
+            for (int j = 0; j < 8; j++) {
+                if (board[i][j] == king) {
+                    kingRow = i;
+                    kingCol = j;
+                    break;
+                }
+            }
+        }
```
Сканить всю доску в поиске короля на каждый ходе совсем не оптимально. Я бы посоветовал в начале игры запоминать где то в свойстве класса позиции королей, и изменять их, когда мы понимаем что фигура, которая делает ход - король. Тогда отпадет нужда сканить доску

trimanul commented on exercises/Chess-build/main.cpp:
```diff
@@ -0,0 +1,132 @@
+            return false;
+        }
+
+        int fromCol = tolower(move[3]) - 'a';
+        int fromRow = move[4] - '1';
+        int toCol = tolower(move[6]) - 'a';
+        int toRow = move[7] - '1';
```
Я бы посоветовал вынести эти операции в 2 функции - calcRowDiff и calcColDiff, т.к. сейчас это просто вычитание char из char, и зачем оно нужно - с первого взгляда неочевидно

trimanul commented on exercises/Chess-build/main.cpp:
```diff
@@ -0,0 +1,132 @@
+            }
+        }
+
+        if (CheckValidator::isKingInCheck(king, kingRow, kingCol, board)) {
+            cout << "Король находится под шахом! Ход отменен." << endl;
+            board[fromRow][fromCol] = piece;
+            board[toRow][toCol] = EMPTY;
+            return false;
+        }
```
Если я правильно уловил логику данной проверки и функции isKingInCheck - то мы не можем походить никакой фигурой кроме короля, если он в шахе. Но нет никакой проверки на мат - необходимо проверять, что у короля больше нет валидных мувов, выводить сообщение о гг и завершать игру.

tootiltedboy committed:
feat(chess): finish console chess gameplay

- replace partial validators with full legal move generation and game state tracking\n- add check, checkmate, stalemate, castling, en passant and promotion\n- improve CLI parsing, board redraw and move feedback\n- make the Chess CMake target use portable project-relative paths

trimanul reviewed (approved):
slop approved
